backend/koofr/koofr.go

```diff
@@ -0,0 +1,461 @@
+package koofr
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/ncw/rclone/fs"
+	"github.com/ncw/rclone/fs/config/configmap"
+	"github.com/ncw/rclone/fs/config/configstruct"
+	"github.com/ncw/rclone/fs/config/obscure"
+	"github.com/ncw/rclone/fs/hash"
+
+	httpclient "github.com/koofr/go-httpclient"
+	koofrclient "github.com/koofr/go-koofrclient"
+)
+
+func init() {
+	fs.Register(&fs.RegInfo{
+		Name:        "koofr",
+		Description: "Koofr",
+		NewFs:       NewFs,
+		Options: []fs.Option{
+			{
+				Name:     "baseurl",
+				Help:     "Base URL of the Koofr API to connect to",
+				Default:  "https://app.koofr.net",
+				Required: true,
+				Advanced: true,
+			}, {
+				Name:     "mountid",
+				Help:     "Mount ID of the mount to use. If omitted, the primary mount is used.",
+				Required: false,
+				Default:  "",
+				Advanced: true,
+			}, {
+				Name:     "user",
+				Help:     "Your Koofr user name",
+				Required: true,
+			}, {
+				Name:       "password",
+				Help:       "Your Koofr password for rclone (generate one at https://app.koofr.net/app/admin/preferences/password)",
+				IsPassword: true,
+				Required:   true,
+			},
+		},
+	})
+}
+
+type options struct {
+	BaseURL  string `config:"baseurl"`
+	MountID  string `config:"mountid"`
+	User     string `config:"user"`
+	Password string `config:"password"`
+}
+
+type koofrfs struct {
+	name     string
+	mountID  string
+	root     string
+	opt      options
+	features *fs.Features
+	client   *koofrclient.KoofrClient
+}
+
+type object struct {
+	fs     *koofrfs
+	remote string
+	info   koofrclient.FileInfo
+}
+
+func base(pth string) string {
+	rv := path.Base(pth)
+	if rv == "" || rv == "." {
+		rv = "/"
+	}
+	return rv
+}
+
+func dir(pth string) string {
+	rv := path.Dir(pth)
+	if rv == "" || rv == "." {
+		rv = "/"
+	}
+	return rv
+}
+
+func (o *object) String() string {
+	return o.remote
+}
+
+func (o *object) Remote() string {
+	return o.remote
+}
+
+func (o *object) ModTime() time.Time {
+	return time.Unix(o.info.Modified/1000, (o.info.Modified%1000)*1000*1000)
+}
+
+func (o *object) Size() int64 {
+	return o.info.Size
+}
+
+func (o *object) Fs() fs.Info {
+	return o.fs
+}
+
+func (o *object) Hash(typ hash.Type) (string, error) {
+	if typ == hash.MD5 {
+		return o.info.Hash, nil
+	}
+	return "", nil
+}
+
+func (o *object) fullPath() string {
+	return o.fs.fullPath(o.remote)
+}
+
+func (o *object) Storable() bool {
+	return true
+}
+
+func (o *object) SetModTime(mtime time.Time) error {
+	return nil
```

Comment:
It is a shame koofr can't set the modified time of a file.

Is there metadata that rclone could use to save the modified time?

Reply:
arbitrary metadata is only available on files stored on our servers, but since we also allow acces to some third-party cloud storage providers as well as files from users machines via the same API, this would not be a universal solution. we'll look into what can be done about this on our side.